hpx/util/detail/pack_traversal_async_impl.hpp

```diff
@@ -0,0 +1,574 @@
+                ++other;
+                return other;
+            }
+
+            bool is_finished() const
+            {
+                return begin_ == sentinel_;
+            }
+        };
+
+        template <typename T>
+        using dynamic_async_range_of_t = dynamic_async_range<
+            typename std::decay<decltype(std::begin(std::declval<T>()))>::type,
+            typename std::decay<decltype(std::end(std::declval<T>()))>::type>;
+
+        /// Returns a dynamic range for the given type
+        template <typename T, typename Range = dynamic_async_range_of_t<T>>
+        Range make_dynamic_async_range(T&& element)
+        {
+            return Range{std::begin(element), std::end(element)};
+        }
+
+        /// Represents a particular point in a asynchronous traversal hierarchy
+        template <typename Frame, typename... Hierarchy>
+        class async_traversal_point
+        {
+            Frame frame_;
+            tuple<Hierarchy...> hierarchy_;
+            bool detached_;
+
+        public:
+            explicit async_traversal_point(
+                Frame frame, tuple<Hierarchy...> hierarchy)
+              : frame_(std::move(frame))
+              , hierarchy_(std::move(hierarchy))
+              , detached_(false)
+            {
+            }
+
+            // Abort the current control flow
+            void detach() noexcept
+            {
+                HPX_ASSERT(!detached_);
+                detached_ = true;
+            }
+
+            /// Returns true when we should abort the current control flow
+            bool is_detached() const noexcept
+            {
+                return detached_;
+            }
+
+            /// Creates a new traversal point which
+            template <typename Parent>
+            auto push(Parent&& parent) -> async_traversal_point<Frame,
+                typename std::decay<Parent>::type, Hierarchy...>
+            {
+                // Create a new hierarchy which contains the
+                // the parent (the last traversed element).
+                auto hierarchy = util::tuple_cat(
+                    util::make_tuple(std::forward<Parent>(parent)), hierarchy_);
+
+                return async_traversal_point<Frame,
+                    typename std::decay<Parent>::type, Hierarchy...>(
+                    frame_, std::move(hierarchy));
+            }
+
+            /// Forks the current traversal point and continues the child
+            /// of the given parent.
+            template <typename Child, typename Parent>
+            void fork(Child&& child, Parent&& parent)
+            {
+                // Push the parent on top of the hierarchy
+                auto point = push(std::forward<Parent>(parent));
+
+                // Continue the traversal with the current element
+                point.async_traverse(std::forward<Child>(child));
+
+                // Propagate the control detach back
+                if (point.is_detached())
```

Comment:
How can a newly created frame be in detached state?

Reply:
This isn't related to a future frame.

The control here is detached if the visited child element detached the control flow.
This happens for instance if we visit a sequence of futures:
```cpp
tuple<future<void>, std::vector<future<int>>>
```
and reach the child `std::vector<future<int>>` of the `tuple` and one of it's containing `futures` isn't ready and we have to wait for its completion asynchronously.

Reply:
Ah, you are correct, I missed that we are doing a `async_traverse` before checking, thanks for the clarification.